static/js/lib/ckeditor/turndown.ts

```diff
@@ -111,3 +111,121 @@ itemRule.replacement = (
+ *
+ * Issue #1
+ * -----------
+ * The default behavior can result in HTML tags beginning a markdown line that
+ * would otherwise be a Markdown paragraph. For example:
+ * ```js
+ * const tds = new TurndownService()
+ * tds.keep(["sup"])
+ * tds.turndown("<p><sup>2</sup> Hello <strong>world</strong>!</p>")
+ * // => "<sup>2</sup> Hello **world**!"
+ * ```
+ * The output above is bad because when an HTML tag begins a markdown line, it
+ * begins a *block* of markdown, and subsequent characters on that line are
+ * NOT treated as Markdown. Hence the `**` above would be rendered as literal
+ * asterisks, not as bold.
+ *
+ * Issue #2
+ * ------------
+ * The default behavior is that the *content* of an HTML tag is not converted
+ * to Markdown. This is problematic for tags that contain shortcode HTML. For
+ * example, `<sup><a class="resource-link" data-uuid ... >2</a></sup>` should
+ * really needs its content to be converted to markdown in order for Hugo to
+ * recognize it as a shortcode.
+ */
+class TurndownHtmlHelpers {
+  private turndownInstance: Turndown
+
+  constructor(turndown: Turndown) {
+    this.turndownInstance = turndown
+  }
+
+  /**
+   * This function improves upon Turndown's default rule for including raw HTML
+   * tags in Markdown. In particular, it:
+   *  - ensures that the child content of an inline HTML node is also converted, which
+   *    we need in case the child content contains shortcode.
+   *  - ensures that an inline HTML tag never begins a line of Markdown.
+   *
+   * Should be used alongside {@link TurndownHtmlHelpers.turndown}
+   */
+  keepReplacer: Turndown.ReplacementFunction = (content, node) => {
+    /**
+     * "isBlock" is a Turndown-specific addition to the DOM node interface.
+     * It appears to be part of their public API in that the author recommends
+     * using it in several issues.
+     */
+    const el = node as HTMLElement & { isBlock: boolean }
+
+    if (el.isBlock) {
+      throw new Error("Inclusion of block content not yet supported.")
+    }
+
+    /**
+     * Now convert the node's child content to Markdown.
+     * This addresses Issue #2 above.
+     */
+    const clone = el.cloneNode() as HTMLElement
+    clone.innerHTML = this.turndown(el.innerHTML)
+
+    /**
+     * To address Issue #1, we'll mark places where HTML is inserted. Later, we
+     * can remove all the markings except the ones that begin a new line.
+     */
+    return `<raw_inline_html/>${clone.outerHTML}`
+  }
+
+  /**
+   * Convert HTML to markdown and insert a zero-width space before any HTML
+   * tag that begins what would otherwise be a Markdown paragraph. Should
+   * be used alongside {@link TurndownHtmlHelpers.keepReplacer}.
+   *
+   * See {@link TurndownHtmlHelpers} for more details.
+   */
+  turndown = (html: string) =>
+    this.turndownInstance
+      .turndown(html)
+      .replace(/^[ ]*<raw_inline_html\/>/g, "\u200b")
+      .replace(/<raw_inline_html\/>/g, "")
```

Comment:
Markdown treats HTML tags at the beginning of a line and elsewhere in a line very differently: html at the beginning starts an [*html block*](https://spec.commonmark.org/0.30/#html-blocks), which disables markdown for the remainder of that line. HTML tags within the line form [inline html](https://spec.commonmark.org/0.30/#raw-html), which allows markdown syntax elsewhere in the line and within the html content.

So if we have an inline html element that would start a line, insert a zero-width space so that it doesn't actually start the line.

I am fairly happy with the whole HTML tag approach to subsup, except for this aspect. And while subscripts at the beginning of the line probably won't occur, superscripts could, e.g., for footnotes[^1].

[^1]: although there are much better ways to do footnotes in markdown, though we don't support them right now.